extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/dart/controller/DartTableInputSpecSlicer.java

```diff
@@ -0,0 +1,290 @@
+
+    for (int i = 0; i < workerIds.size(); i++) {
+      reverseWorkers.put(WorkerId.fromString(workerIds.get(i)).getHostAndPort(), i);
+    }
+
+    return new DartTableInputSpecSlicer(reverseWorkers, serverView);
+  }
+
+  @Override
+  public boolean canSliceDynamic(final InputSpec inputSpec)
+  {
+    return false;
+  }
+
+  @Override
+  public List<InputSlice> sliceStatic(final InputSpec inputSpec, final int maxNumSlices)
+  {
+    final TableInputSpec tableInputSpec = (TableInputSpec) inputSpec;
+    final TimelineLookup<String, ServerSelector> timeline =
+        serverView.getTimeline(new TableDataSource(tableInputSpec.getDataSource()).getAnalysis()).orElse(null);
+
+    if (timeline == null) {
+      return Collections.emptyList();
+    }
+
+    final Set<QueryableDataSegment> prunedSegments =
+        findQueryableDataSegments(
+            tableInputSpec,
+            timeline,
+            serverSelector -> findWorkerForServerSelector(serverSelector, maxNumSlices)
+        );
+
+    final List<List<QueryableDataSegment>> assignments = new ArrayList<>(maxNumSlices);
+    while (assignments.size() < maxNumSlices) {
+      assignments.add(null);
+    }
+
+    int nextRoundRobinWorker = 0;
+    for (final QueryableDataSegment segment : prunedSegments) {
+      final int worker;
+      if (segment.getWorkerNumber() == UNKNOWN) {
+        // Segment is not available on any worker. Assign to some worker, round-robin. Today, that server will throw
+        // an error about the segment not being findable, but perhaps one day, it will be able to load the segment
+        // on demand.
+        worker = nextRoundRobinWorker;
+        nextRoundRobinWorker = (nextRoundRobinWorker + 1) % maxNumSlices;
+      } else {
+        worker = segment.getWorkerNumber();
+      }
+
+      if (assignments.get(worker) == null) {
+        assignments.set(worker, new ArrayList<>());
+      }
+
+      assignments.get(worker).add(segment);
+    }
+
+    return makeSegmentSlices(tableInputSpec.getDataSource(), assignments);
+  }
+
+  @Override
+  public List<InputSlice> sliceDynamic(
+      final InputSpec inputSpec,
+      final int maxNumSlices,
+      final int maxFilesPerSlice,
+      final long maxBytesPerSlice
+  )
+  {
+    throw new UnsupportedOperationException();
+  }
+
+  /**
+   * Return the worker ID that corresponds to a particular {@link ServerSelector}, or {@link #UNKNOWN} if none does.
+   *
+   * @param serverSelector the server selector
+   * @param maxNumSlices   maximum number of worker IDs to use
+   */
+  int findWorkerForServerSelector(final ServerSelector serverSelector, final int maxNumSlices)
+  {
+    final QueryableDruidServer<?> server = serverSelector.pick(null);
+
+    if (server == null) {
+      return UNKNOWN;
+    }
+
+    final String serverHostAndPort = server.getServer().getHostAndPort();
+    final int worker = reverseWorkers.getInt(serverHostAndPort);
+
+    if (worker != UNKNOWN && worker < maxNumSlices) {
```

Comment:
I don't think it's possible. `UNKNOWN` is possible in case of race condition though. I added this comment:

```
    // The worker number may be UNKNOWN in a race condition, such as the set of Historicals changing while
    // the query is being planned. I don't think it can be >= maxNumSlices, but if it is, treat it like UNKNOWN.
```